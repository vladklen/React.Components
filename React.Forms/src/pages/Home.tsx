import { useSelector } from 'react-redux';
import { DataType } from '../types/types';
import { IFormData } from '../store/Slices/formData.slice';
import { useEffect, useState } from 'react';

const Home = () => {
  const { form } = useSelector((state: DataType) => state);
  const [active, setActive] = useState(false);

  useEffect(() => {
    setActive(true);
    const timer = setTimeout(() => setActive(false), 2000);
    return () => clearTimeout(timer);
  }, []);

  return (
    <>
      <h1>Main page</h1>
      {[...form].reverse().map((el: IFormData, index) => {
        return (
          <div key={index} className={index === 0 && active ? 'active' : ''}>
            {Object.entries(el).map(([key, value]) => {
              if (key === 'image') {
                return <img key={key} src={value} id="base64image" />;
              }
              if (key === 'accept') {
                return null;
              }
              return (
                <p key={key}>
                  {key}:{value}
                </p>
              );
            })}
          </div>
        );
      })}
    </>
  );
};

export default Home;
