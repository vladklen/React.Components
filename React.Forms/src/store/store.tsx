import { configureStore } from '@reduxjs/toolkit';
import formDataReducer from './Slices/formData.slice';
import countryDataReducer from './Slices/countryData.slice';

const store = configureStore({
  reducer: {
    form: formDataReducer,
    countries: countryDataReducer,
  },
});
export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export default store;
