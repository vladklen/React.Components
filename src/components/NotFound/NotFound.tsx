import Link from 'next/link';

export default function NotFound() {
  return (
    <div>
      <h2>Page not found!</h2>
      <p>
        Go to the<Link href="/">Home</Link>
      </p>
    </div>
  );
}
