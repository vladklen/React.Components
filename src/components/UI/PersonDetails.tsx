import { useNavigate, useParams } from 'react-router-dom';
import { useEffect, useState } from 'react';
import { ColorRing } from 'react-loader-spinner';
import { ModalWrapper, StyledCard } from './Styles';

// interface CardProps {
//   birth_year: string;
//   name: string;
//   height: string;
//   mass: string;
//   gender: string;
// }

export default function PersonDetails() {
  let navigate = useNavigate();
  const { name } = useParams();
  const [loading, setLoading] = useState(false);
  const [content, setContent] = useState([]);

  useEffect(() => {
    const startSearch = async () => {
      setLoading(true);
      const response = await fetch(
        `https://swapi.dev/api/people/?search=${name}`
      );
      const data = await response.json();
      setLoading(false);
      setContent(data.results);
    };
    startSearch();
  }, [name]);

  return (
    <ModalWrapper onClick={() => navigate(-1)}>
      <StyledCard>
        {content.length && !loading ? (
          <h3>Name: {content[0].name}</h3>
        ) : (
          <ColorRing
            visible
            height="80"
            width="80"
            ariaLabel="blocks-loading"
            wrapperStyle={{}}
            wrapperClass="blocks-wrapper"
            colors={['#e15b64', '#f47e60', '#f8b26a', '#abbd81', '#849b87']}
          />
        )}
      </StyledCard>
    </ModalWrapper>
  );
}
