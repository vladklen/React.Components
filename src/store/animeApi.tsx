import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';
import baseAPI from '../constants/api';
import { IAnime, IDataResponse } from '../types/types';

export const animeApi = createApi({
  reducerPath: 'animeApi',
  baseQuery: fetchBaseQuery({ baseUrl: `${baseAPI}/anime` }),
  endpoints: (builder) => ({
    getCardList: builder.query<
      IDataResponse,
      { search: string; page: string; limit: string }
    >({
      query: ({ search, page, limit }) => {
        return {
          url: `?q=${search}&page=${page}&limit=${limit}`,
        };
      },
    }),
    getCardById: builder.query<IAnime, string>({
      query: (id) => {
        return {
          url: `/${id}`,
        };
      },
      transformResponse: (response: { data: IAnime }) => response.data,
    }),
  }),
});

export const { useGetCardListQuery, useGetCardByIdQuery } = animeApi;
