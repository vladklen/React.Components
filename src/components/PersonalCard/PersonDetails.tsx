import { useNavigate, useParams } from 'react-router-dom';
import { useEffect, useState } from 'react';
import { ColorRing } from 'react-loader-spinner';
import { ModalWrapper } from '../UI/Styles';
import { IAnime, getAnimeById } from '../../api/StartSearch';
import MyButton from '../UI/MyButton/MyButton';
import {
  StyledPersonalCard,
  StyledPersonalCardContent,
  StyledCardImage,
} from './Styles';

export default function PersonDetails() {
  const navigate = useNavigate();
  const { id } = useParams();
  const [loading, setLoading] = useState(false);
  const [content, setContent] = useState<IAnime | null>(null);

  useEffect(() => {
    if (id) {
      const startSearch = async () => {
        setLoading(true);
        const data = await getAnimeById(id);
        setLoading(false);
        if (data) {
          setContent(data.data);
        }
      };
      startSearch();
    }
  }, [id]);

  return (
    <ModalWrapper onClick={() => navigate(-1)} data-testid={`cardDetails${id}`}>
      <StyledPersonalCard>
        {content && !loading ? (
          <>
            <StyledCardImage
              style={{
                backgroundImage: `url("${content.images.jpg.image_url}")`,
              }}
            />
            <StyledPersonalCardContent>
              <h3>Title: {content.title}</h3>
              <h3>{content.title_japanese}</h3>
              <p>Rating:{content.rating}</p>
              <p>Status:{content.status}</p>
              <p>Score: {content.score}</p>
              <MyButton
                click={() => navigate(-1)}
                color="blue"
                message="Close"
                dataTest="test-CloseButton"
              />
            </StyledPersonalCardContent>
          </>
        ) : (
          <div data-testid="test-loader">
            <ColorRing
              data-testid="test-cardDetails-loader"
              visible
              height="80"
              width="80"
              ariaLabel="blocks-loading"
              wrapperStyle={{}}
              wrapperClass="blocks-wrapper"
              colors={['#e15b64', '#f47e60', '#f8b26a', '#abbd81', '#849b87']}
            />
          </div>
        )}
      </StyledPersonalCard>
    </ModalWrapper>
  );
}
