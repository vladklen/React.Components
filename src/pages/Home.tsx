/* eslint-disable no-restricted-syntax */
import { useContext, useEffect, useState } from 'react';
import { ColorRing } from 'react-loader-spinner';
import { Outlet, useLocation, useSearchParams } from 'react-router-dom';
import { ContentWrapper, SearchWrapper } from '../components/UI/Styles';
import MyInput from '../components/UI/MyInput/MyInput';
import MyButton from '../components/UI/MyButton/MyButton';
import Pagination from '../components/UI/Pagination/Pagination';
import SelectAmount from '../components/UI/SelectAmount';
import { getAnime } from '../api/StartSearch';
import { AppContext } from '../context/Context';
import CardList from '../components/CardList/CardList';

export default function Home() {
  const { value, setValue, data, setData } = useContext(AppContext);
  const [search, setSearch] = useSearchParams();
  const [totalPosts, setTotalPosts] = useState(0);
  const [postsPerPage, setPostsPerPage] = useState(10);
  const [loading, setLoading] = useState(false);
  const location = useLocation();

  const handleInputChange = (event: { target: { value: string } }) => {
    setValue(event.target.value);
    localStorage.setItem('test', event.target.value);
  };

  const handleSearchStart = () => {
    let params = {};
    for (const [key, val] of search) {
      params = { ...params, [key]: val, page: '1', search: value };
    }
    setSearch(params);
  };

  useEffect(() => {
    const startSearch = async () => {
      setLoading(true);
      const response = await getAnime(search);
      setData(response.data);
      setPostsPerPage(response.pagination.items.per_page);
      setLoading(false);
      setTotalPosts(response.pagination.items.total);
    };

    if (location.pathname.length === 1) {
      if (
        !search.get('search') &&
        !search.get('page') &&
        !search.get('limit')
      ) {
        if (localStorage.getItem('test')) {
          setSearch({
            search: String(localStorage.getItem('test')),
            page: '1',
            limit: '10',
          });
        } else {
          setSearch({ page: '1', limit: '10' });
        }
        return;
      }
      if (search.get('limit')) {
        let params = {};
        let limit: number = Number(search.get('limit'));
        if (limit > 5) {
          limit = 10;
        } else if (limit > 1) {
          limit = 5;
        } else {
          limit = 1;
        }
        for (const [key, val] of search) {
          params = { ...params, [key]: val, limit: `${limit}` };
        }
        setSearch(params);
      }
      startSearch();
    }
  }, [location.pathname.length, search, setData, setSearch]);

  return (
    <div>
      <SearchWrapper>
        <MyInput change={handleInputChange} />
        <MyButton click={handleSearchStart} color="blue" message="Search" />
      </SearchWrapper>
      <h2>Results:</h2>
      <ContentWrapper>
        {loading && (
          <ColorRing
            visible
            height="80"
            width="80"
            ariaLabel="blocks-loading"
            wrapperStyle={{}}
            wrapperClass="blocks-wrapper"
            colors={['#e15b64', '#f47e60', '#f8b26a', '#abbd81', '#849b87']}
          />
        )}
        {!loading && <CardList />}
        <Outlet />
      </ContentWrapper>
      <SelectAmount />
      <Pagination
        postsPerPage={postsPerPage}
        totalPosts={totalPosts}
        loading={loading}
      />
    </div>
  );
}
