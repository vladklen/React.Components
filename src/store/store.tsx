import { combineReducers, configureStore } from '@reduxjs/toolkit';
import searchReducer from './searchValue/searchValue.slice';
import postsReducer from './postsPerPage/postsPerPage.slice';
import { animeApi } from './animeApi';

export const rootReducer = combineReducers({
  value: searchReducer,
  postsPerPage: postsReducer,
  [animeApi.reducerPath]: animeApi.reducer,
});

const store = configureStore({
  reducer: rootReducer,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(animeApi.middleware),
  devTools: true,
});
export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export default store;
