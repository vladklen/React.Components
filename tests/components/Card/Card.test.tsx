import { describe, expect, test, vi } from 'vitest';
import { act, fireEvent, render } from '@testing-library/react';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import Home from '../../../src/pages/Home';
import { data, pagination } from '../../mocks/AnimeRespone';
import PersonDetails from '../../../src/components/PersonalCard/PersonDetails';
import NotFound from '../../../src/pages/NotFound';
import { useGetCardByIdQuery } from '../../../src/store/animeApi';

const mockState = {
  value: {
    search: 'test',
    page: '1',
    limit: '10',
  },
};

vi.mock('react-redux', () => ({
  useDispatch: () => vi.fn(),
  useSelector: vi.fn(() => mockState),
}));

vi.mock('../../../src/store/animeApi', () => ({
  useGetCardListQuery: vi.fn(() => ({
    isLoading: false,
    data: { data, pagination },
  })),
  useGetCardByIdQuery: vi.fn(() => ({ isLoading: false, data: data[0] })),
}));

describe('Test Card component', () => {
  test('Test Home component render', async () => {
    const wrapper = render(
      <MemoryRouter>
        <Home />
      </MemoryRouter>
    );

    const items = await wrapper.findAllByRole('link');

    expect(items.length).toBe(3);
  });

  test('Ensure that the card component renders the relevant card data;', async () => {
    const wrapper = render(
      <MemoryRouter>
        <Home />
      </MemoryRouter>
    );

    const items = await wrapper.findAllByRole('link');

    expect(items[0]).toMatchSnapshot();
  });

  test('Validate that clicking on a card opens a detailed card component', async () => {
    const wrapper = render(
      <MemoryRouter initialEntries={['/']}>
        <Routes>
          <Route path="/" element={<Home />}>
            <Route path="details/:id" element={<PersonDetails />} />
          </Route>
          <Route path="*" element={<NotFound />} />
        </Routes>
      </MemoryRouter>
    );
    const linkElement = await wrapper.findByTestId(`test${data[0].mal_id}`);

    act(() => {
      fireEvent.click(linkElement);
    });

    expect(
      await wrapper.findByTestId(`cardDetails${data[0].mal_id}`)
    ).toBeTruthy();
  });

  test('Check that clicking triggers an additional API call to fetch detailed information', async () => {
    const wrapper = render(
      <MemoryRouter initialEntries={['/']}>
        <Routes>
          <Route path="/" element={<Home />}>
            <Route path="details/:id" element={<PersonDetails />} />
          </Route>
          <Route path="*" element={<NotFound />} />
        </Routes>
      </MemoryRouter>
    );
    const linkElement = await wrapper.findByTestId(`test${data[0].mal_id}`);

    act(() => {
      fireEvent.click(linkElement);
    });

    expect(useGetCardByIdQuery).toBeCalled();
  });
});
