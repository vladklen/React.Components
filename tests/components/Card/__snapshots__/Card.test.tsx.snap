// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test CardList component > Ensure that the card component renders the relevant card data; 1`] = `
<a
  data-testid="test1"
  href="/details/1"
>
  <div
    class="sc-iCoHzw FnKDP"
  >
    <h3>
      Test title #1
    </h3>
    <div
      class="sc-fujznN epBsNf"
      style="background-image: url(url:page1);"
    />
  </div>
</a>
`;

exports[`Test CardList component > Validate that clicking on a card opens a detailed card component 1`] = `
<div
  class="sc-dlniIP fnahtJ"
  data-testid="cardDetails1"
>
  <div
    class="sc-pNWRh joCXoT"
  >
    <div
      class="sc-kEqXeH jDIRYD"
      style="background-image: url(url:page1);"
    />
    <div
      class="sc-jrsKJM cDpoem"
    >
      <h3>
        Title: 
        Test title #1
      </h3>
      <h3>
        Title in Japanese #1
      </h3>
      <p>
        Rating:
        1
      </p>
      <p>
        Status:
        1
      </p>
      <p>
        Score: 
        1
      </p>
      <button
        class="sc-jSFipO eTGvZn"
        type="button"
      >
        Close
         
      </button>
    </div>
  </div>
</div>
`;
